File: src/templates/tmpl.erl
% Erlang
% escript skips line 1 (reserved for a shebang), so start with a blank line
<QCConst>newline</QCConst>
main(V)->io:setopts([{encoding,latin1}]),
  lists:foldl(fun(G,C)->put(C,G),C+1 end,0,hd(V++["erl"])++[0,0,0,0,0,0,0,0]),
  s(<QCConst>entrypoint</QCConst>,0,0,[]).<QCConst>newline</QCConst>

% ---- read: a sentinel 0 at the end removes the range guard (at most one read past the end) ----
b(J)->binary:at(<<"<QCImage/>",0>>,J).<QCConst>newline</QCConst>

% ---- literal output loop (the Hx -> x-33 escape happens only here) and ( ) skipping ----
li(K)->C=b(K),if C==47->K+1;C==72->io:put_chars([b(K+1)-33]),li(K+2);true->io:put_chars([C]),li(K+1) end.<QCConst>newline</QCConst>
sk(K,E)->C=b(K+1),R=if C>39,C<42->E+81-2*C;true->E end,if R>0->sk(K+1,R);true->K+1 end.<QCConst>newline</QCConst>

% ---- main loop (tail recursive; branches on ranges of C with if) ----
s(K,P,A,L)->if K<<QCConst>image_len</QCConst>->C=b(K),J=K+1,
    if C<34->s(li(J),P,A,L);
      C<39->s(J,P+1,b(P),L);
      C<41->s(sk(J,1),P,A,[J|L]);
      C<42->if A==0->s(J,P,A,tl(L));true->s(hd(L),P,A,L) end;
      C<43->io:put_chars([A]),s(J,P,A,L);
      C<56->put(C rem 8,A),s(J,P,A,L);
      C<64->s(J,P,get(C rem 8),L);
      C<72->s(J,P,A-get(C rem 8),L);
      true->s(J,P,C,L) end;
    true->ok end.<QCConst>newline</QCConst>

File: src/templates/tmpl.exs
# Elixir

import Process,only: [put: 2,get: 2];
defmodule Q do

  # ---- a sentinel 0 at the end removes the range guard (at most one read past the end) ----
  def b(j),do: :binary.at(<< ~S"<QCImage/>",0>>,j);

  def o(x),do: :io.put_chars([x]);

  # ---- literal output loop (the Hx -> x-33 escape happens only here) and ( ) skipping ----
  def li(i) do c=b(i);cond do c==47->i+1;c==72->o(b(i+1)-33);li(i+2);true->o(c);li(i+1) end end;
  def sk(i,d) do c=b(i+1);d=if c>39 and c<42,do: d+81-2*c,else: d;if d>0,do: sk(i+1,d),else: i+1 end;

  # ---- main loop (tail recursive; branches on ranges of c with cond) ----
  def s(i,p,a,f) do if i<<QCConst>image_len</QCConst> do c=b(i);i=i+1;cond do c<34->s(li(i),p,a,f);
    c<39->s(i,p+1,b(p),f);
    c<41->s(sk(i,1),p,a,[i|f]);
    c<42->if a==0,do: s(i,p,a,tl(f)),else: s(hd(f),p,a,f);
    c<43->o(a);s(i,p,a,f);
    c<56->put(rem(c,8),a);s(i,p,a,f);
    c<64->s(i,p,get(rem(c,8),0),f);
    c<72->s(i,p,a-get(rem(c,8),0),f);
    true->s(i,p,c,f);
  end end end end;

# ---- start: put the name into r[0..] and run from the entrypoint ----
:io.setopts([{:encoding,:latin1}]);
Enum.reduce(to_charlist(hd(System.argv++["exs"])),0,fn x,c->put(c,x);c+1 end);
Q.s(<QCConst>entrypoint</QCConst>,0,0,[])
